drivers/azure/azure.go

```diff
@@ -0,0 +1,455 @@
+		func() error { return c.DeleteVirtualMachineIfExists(d.ResourceGroup, d.naming().VM()) },
+		func() error { return c.DeleteNetworkInterfaceIfExists(d.ResourceGroup, d.naming().NIC()) },
+		func() error { return c.DeletePublicIPAddressIfExists(d.ResourceGroup, d.naming().IP()) },
+		func() error { return c.DeleteNetworkSecurityGroupIfExists(d.ResourceGroup, d.naming().NSG()) },
+		func() error { return c.CleanupAvailabilitySetIfExists(d.ResourceGroup, d.AvailabilitySet) },
+		func() error { return c.CleanupSubnetIfExists(d.ResourceGroup, d.VirtualNetwork, d.SubnetName) },
+		func() error { return c.CleanupVirtualNetworkIfExists(d.ResourceGroup, d.VirtualNetwork) },
+	} {
+		if err := f(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetIP returns public IP address or hostname of the machine instance.
+func (d *Driver) GetIP() (string, error) {
+	if d.resolvedIP == "" {
+		ip, err := d.ipAddress()
+		if err != nil {
+			return "", err
+		}
+		d.resolvedIP = ip
+	}
+	log.Debugf("Machine IP address resolved to: %s", d.resolvedIP)
+	return d.resolvedIP, nil
+}
+
+// GetSSHHostname returns an IP address or hostname for the machine instance.
+func (d *Driver) GetSSHHostname() (string, error) {
+	return d.GetIP()
+}
+
+// GetURL returns a socket address to connect to Docker engine of the machine
+// instance.
+func (d *Driver) GetURL() (string, error) {
+	if err := drivers.MustBeRunning(d); err != nil {
+		return "", err
+	}
+
+	// NOTE (ahmetalpbalkan) I noticed that this is not used until machine is
+	// actually created and provisioned. By then GetIP() should be returning
+	// a non-empty IP address as the VM is already allocated and connected to.
+	ip, err := d.GetIP()
+	if err != nil {
+		return "", err
+	}
+	u := (&url.URL{
+		Scheme: "tcp",
+		Host:   net.JoinHostPort(ip, fmt.Sprintf("%d", d.DockerPort)),
+	}).String()
+	log.Debugf("Machine URL is resolved to: %s", u)
+	return u, nil
+}
+
+// GetState returns the state of the virtual machine role instance.
+func (d *Driver) GetState() (state.State, error) {
+	c, err := d.newAzureClient()
+	if err != nil {
+		return state.None, err
+	}
+	powerState, err := c.GetVirtualMachinePowerState(
+		d.ResourceGroup, d.naming().VM())
+	if err != nil {
+		return state.None, err
+	}
+
+	machineState := machineStateForVMPowerState(powerState)
+	log.Debugf("Determined Azure PowerState=%q, docker-machine state=%q",
+		powerState, machineState)
+	return machineState, nil
+}
+
+// Start issues a power on for the virtual machine instance.
+func (d *Driver) Start() error {
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+	return c.StartVirtualMachine(d.ResourceGroup, d.naming().VM())
+}
+
+// Stop issues a power off for the virtual machine instance.
+func (d *Driver) Stop() error {
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+	log.Info("NOTICE: Stopping an Azure Virtual Machine is just going to power it off, not deallocate.")
```

Comment:
@nathanleclaire hmm, so deallocationg will stop charging. Stopping on the other hand, will keep charging but will keep the IP address the same. (for instance if you stop/start google machines, you run into cert issues because IP address is ephemeral; not static, and I took the same approach here). Also starting deallocated VM will be slightly slower to start; starting a stopped VM on the other hand is way faster.

I'm open to any suggestions about what drivers should do in this case.